File: rust/features/steps/output_steps.rs
use std::sync::OnceLock;

use cucumber::then;
use regex::Regex;

use crate::step_definitions::initialization_steps::TaskulusWorld;

#[then(expr = "stdout should contain {string}")]
fn then_stdout_contains_text(world: &mut TaskulusWorld, text: String) {
    let stdout = strip_ansi(world.stdout.as_ref().expect("stdout"));
    let normalized = text.replace("\\\"", "\"");
    assert!(stdout.contains(&normalized));
}

#[then(expr = "stdout should not contain {string}")]
fn then_stdout_not_contains_text(world: &mut TaskulusWorld, text: String) {
    let stdout = strip_ansi(world.stdout.as_ref().expect("stdout"));
    let normalized = text.replace("\\\"", "\"");
    assert!(!stdout.contains(&normalized));
}

#[then(expr = "stderr should contain {string}")]
fn then_stderr_contains_text(world: &mut TaskulusWorld, text: String) {
    let stderr = strip_ansi(world.stderr.as_ref().expect("stderr"));
    let normalized = text.replace("\\\"", "\"");
    assert!(stderr.contains(&normalized));
}

#[then(expr = "stdout should contain {string} once")]
fn then_stdout_contains_once(world: &mut TaskulusWorld, text: String) {
    let stdout = strip_ansi(world.stdout.as_ref().expect("stdout"));
    let normalized = text.replace("\\\"", "\"");
    assert_eq!(stdout.matches(&normalized).count(), 1);
}

#[then(expr = "stdout should contain the external project path for {string}")]
fn then_stdout_contains_external_project_path(world: &mut TaskulusWorld, identifier: String) {
    let stdout = strip_ansi(world.stdout.as_ref().expect("stdout"));
    let project_path = world
        .expected_project_path
        .as_ref()
        .expect("expected project path");
    let project_path = project_path.to_string_lossy();
    let matches = stdout
        .lines()
        .any(|line| line.contains(identifier.as_str()) && line.contains(project_path.as_ref()));
    assert!(
        matches,
        "no line contains both external project path and identifier"
    );
}

#[then(expr = "stdout should list {string} before {string}")]
fn then_stdout_lists_before(world: &mut TaskulusWorld, first: String, second: String) {
    let stdout = strip_ansi(world.stdout.as_ref().expect("stdout"));
    let first_index = stdout.find(&first).expect("first value not found");
    let second_index = stdout.find(&second).expect("second value not found");
    assert!(first_index < second_index);
}

fn strip_ansi(text: &str) -> String {
    static ANSI_RE: OnceLock<Regex> = OnceLock::new();
    let regex = ANSI_RE.get_or_init(|| Regex::new("\x1b\\[[0-9;]*m").expect("regex"));
    regex.replace_all(text, "").to_string()
}

File: python/features/steps/output_steps.py
"""Behave steps for CLI output assertions."""

from __future__ import annotations

import re

from behave import then

_ANSI_RE = re.compile(r"\x1b\[[0-9;]*m")


def _strip_ansi(text: str) -> str:
    return _ANSI_RE.sub("", text)


@then('stdout should contain "{text}"')
def then_stdout_contains_text(context: object, text: str) -> None:
    normalized = text.replace('\\"', '"')
    stdout = _strip_ansi(context.result.stdout)
    if normalized not in stdout:
        print(f"Expected '{normalized}' to be in stdout, but it wasn't")
        print(f"ACTUAL STDOUT:\n{stdout}")
        print(f"Exit code: {context.result.exit_code}")
        print(f"STDERR:\n{context.result.stderr}")
    assert normalized in stdout


@then('stdout should not contain "{text}"')
def then_stdout_not_contains_text(context: object, text: str) -> None:
    normalized = text.replace('\\"', '"')
    stdout = _strip_ansi(context.result.stdout)
    if normalized in stdout:
        print(f"Expected '{normalized}' to NOT be in stdout, but it was")
        print(f"ACTUAL STDOUT:\n{stdout}")
    assert normalized not in stdout


@then('stderr should contain "{text}"')
def then_stderr_contains_text(context: object, text: str) -> None:
    normalized = text.replace('\\"', '"')
    stderr = _strip_ansi(context.result.stderr)
    assert normalized in stderr


@then('stdout should contain "{text}" once')
def then_stdout_contains_once(context: object, text: str) -> None:
    normalized = text.replace('\\"', '"')
    stdout = _strip_ansi(context.result.stdout)
    assert stdout.count(normalized) == 1


@then('stdout should list "{first}" before "{second}"')
def then_stdout_lists_before(context: object, first: str, second: str) -> None:
    stdout = _strip_ansi(context.result.stdout)
    first_index = stdout.find(first)
    second_index = stdout.find(second)
    assert first_index != -1, f"{first} not found in stdout"
    assert second_index != -1, f"{second} not found in stdout"
    assert first_index < second_index, f"{first} did not appear before {second}"


@then('stdout should contain the external project path for "{identifier}"')
def then_stdout_contains_external_project_path(
    context: object, identifier: str
) -> None:
    project_path = getattr(context, "external_project_path", None)
    assert project_path is not None
    lines = _strip_ansi(context.result.stdout).splitlines()
    matches = [
        line for line in lines if identifier in line and str(project_path) in line
    ]
    assert matches, "no line contains both external project path and identifier"
